### PR TITLE
Sendgrid adr

### LINKED ISSUE #1784: [SPIKE] Backend e-mail relay
**Description:**
_Provide a brief background and justification for this issue_
Investigate usage of `smtp.ees.hhs.gov` as relay from our Cloud.gov spaces.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [x] Produce script(s) or series of commands for sending an e-mail (and email actually gets sent)
- [x] Pro/con list of SMTP tools (python library vs. linux CLI tool vs...)
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] Investigate python libraries for working with smtp
- [x] Investigate smtp cli tools 
- [x] Investigate cloud/webapp smtp tools
- [x] Stretch: proof of concept integration with our Django app
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

**Supporting Documentation:**
- [DKIM hack.md](https://hackmd.io/@dBEtH2T9SRqyVnE3ZKtYSA/rkXq8XPdn)
 - https://github.com/cloud-gov/pages-mailer
 - https://aws.amazon.com/ses/
 - https://docs.python.org/3/library/smtplib.html
 - https://realpython.com/python-send-email/

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- How do we ensure this is secure and couldn't be potentially exploited as an open-source project?

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #1784 

Provides a cost/benefits analysis of various email relay options in the face of ITIO's decision to deprecate connections to their SMTP servers via IronPort.

## How to Test
N/A

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] Create an ADR
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# 1. Record architecture decisions
```
Comment (andrew-jameson) [inline docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md]: This needs to be updated to 21 and relevant title where we typically match the filename.

**Review by andrew-jameson — COMMENTED** [review]
Minor change request. Where is the boundary diagram update for our changes being tracked?

**Comment by jtimpe** [comment]
> Minor change request. Where is the boundary diagram update for our changes being tracked?

@andrew-jameson i was unclear on that - is the boundary diagram something raft devs will update? from the conversation at IPT, it sounded like that was asked of Tijan's team but I'm happy to make an update here as well.

**Review by raftmsohani — APPROVED** [review]

**Comment by andrew-jameson** [comment]
> > Minor change request. Where is the boundary diagram update for our changes being tracked?
> 
> @andrew-jameson i was unclear on that - is the boundary diagram something raft devs will update? from the conversation at IPT, it sounded like that was asked of Tijan's team but I'm happy to make an update here as well.

Yeah, you and I were both unclear. I just didn't want to lose track of that.

The boundary diagram that @ADPennington has updated in the past is [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/Security-Compliance/boundary-diagram.md). To make things more confusing, we have a secondary one [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/Technical-Documentation/diagrams/tdp-environments.png) that devs have updated in the past.

**Review by andrew-jameson — APPROVED** [review]

**Comment by andrew-jameson** [comment]
> > > Minor change request. Where is the boundary diagram update for our changes being tracked?
> > 
> > 
> > @andrew-jameson i was unclear on that - is the boundary diagram something raft devs will update? from the conversation at IPT, it sounded like that was asked of Tijan's team but I'm happy to make an update here as well.
> 
> Yeah, you and I were both unclear. I just didn't want to lose track of that.
> 
> The boundary diagram that @ADPennington has updated in the past is [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/Security-Compliance/boundary-diagram.md). To make things more confusing, we have a secondary one [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/Technical-Documentation/diagrams/tdp-environments.png) that devs have updated in the past.

For posterity, spoke on this and this is fine to merge as the boundary diagram is being tracked elsewhere and will be updated on govt side.

**Comment by andrew-jameson** [comment]
@ADPennington Let me know if you have any objections to the ADR before merging, thanks!

File: docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+> 4. Point your application to the email service provider’s SMTP server(s)
+> 5. Test sending email from your @acf.hhs.gov mail from address(es)
+
+* https://github.com/raft-tech/TANF-app/issues/1784#issuecomment-1613321602
+
+Options
+* Host an SMTP server
+    * Pro: theoretically, we would have more control over our system and security by deploying our own services
+    * Con: this comes with a potentially high maintenance cost
+    * Con: no gloud.gov provided SMTP infrastructure available
```
Comment (ADPennington) [inline docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md]:
```suggestion
    * Con: no cloud.gov provided SMTP infrastructure available
```

File: docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+We will utilize a third-party, non-SMTP email service - [SendGrid](https://sendgrid.com/). SendGrid provides a python library, [`sendgrid-python`](https://github.com/sendgrid/sendgrid-python) that we will implement in a Django `EmailBackend`, allowing a drop-in replacement to the existing `SMTPEmailBackend`. SendGrid also provides more features than similar Email as a Service solutions, such as Amazon Simple Email Service (SES), including deliverability insights and analytics. This allows us to more easily verify and debug issues with email deliverability in production settings.
+
+## Consequences
+
+### Benefits
+
+* Reliable managed email service, less flaky than SMTP
+* Increased visibility vs. SMTP - deliverability insights allow us to view whether or not an email was delivered and reason it bounced back, as well as if it has been opened.
+* Large feature set - retry, analytics, automation, templates; saves us debugging/implementation time by utilizing a well-used, well-documented service.
```
Comment (ADPennington) [inline docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md:49]:
```suggestion
* Large feature set - retry, analytics, automation, templates; saves us debugging/implementation time by utilizing a well-used, well-documented service.
* ACF Tech security team assessed SendGrid along the following dimensions and approved its use for TDP:
  - _compliance:_ SendGrid earned the SOC 2 Type II certification, based on rigorous controls to safeguard customer data
  - _data security:_ access to SendGrid system and data is restricted only to to those who need access in order to provide support. All customer data is encrypted in transit via TLS. 
  - _business continuity/disaster recovery:_ SendGrid has separate data centers to support consistent service delivery for customers in the event of an outage 
  - _privacy:_ SendGrid takes appropriate technical and organizational measures to protect the security of users/ PII both online and offline. For reference, SendGrid only accesses TDP user emails. 
```

File: docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+### Risks
+* Api/library versioning - any breaking changes to the api or python library will have to be prioritized and fixed quickly or risk breaking email notifications for the application.
+* Relying on a third party service - outages, service shutoffs, etc. would affect the deliverability of notification emails and security vulnerabilities could affect user privacy.
+
+## Notes
+
+* [Pro/Con List for SendGrid and Amazon SES](https://hackmd.io/25liUecySPSXOtyiF6YugA)
```
Comment (ADPennington) [inline docs/Technical-Documentation/Architecture-Decision-Record/021-sendgrid.md:64]:
```suggestion
* [link to boundary diagram]()
* [link to technical diagram]()
* [Pro/Con List for SendGrid and Amazon SES](https://hackmd.io/25liUecySPSXOtyiF6YugA)
```

**Review by ADPennington — APPROVED** [review]
👍🏾

**Pull request merged**